crates/zed/src/languages/terraform/config.toml

```diff
@@ -0,0 +1,13 @@
+name = "Terraform"
```

Comment:
I had this thought originally, given there's terraform-specific syntax in the highlighting and the LSP/Formatter is terraform-specific I figured it goes in as Terraform specifically.

I was then torn on adding Nomad, Packer separately or together as HCL as a second language albeit sharing the same tree-sitter-hcl package underneath. Those tend to be named `.nomad.hcl` or `.pkr.hcl` so being defined as HCL together makes sense I think.

Reply:
Indeed, but moving to hcl, would open another question 😅 can we trigger the LSP only on `.tf` and `.tfvars` using the same tree-sitter syntax? I'll try to dig on it between today and tomorrow, so we have all we need for our session

Reply:
Terraform is a more specific language to HCL. HCL: https://developer.hashicorp.com/nomad/docs/job-specification/hcl2/syntax is more generic and is used across different tools from hashicorp and also 3rd party softawre such as Terragrunt.

Terraform files are `.tf` and HCL files are `.hcl`.

Reply:
In my opinion, we should just add a single language, and call it HCL. It's ok if the language server is Terraform specific and there is some Terraform-specific syntax highlighting.

Reply:
We'll (well, I'll) want nomad formatting and packer formatting (which are basically the same I think) which differ to terraform formatting so I think the split in languages makes sense.

HCL is the base language for specifying configuration (and some dynamic parts) in, then Terraform extends that definition with terraform-specific functionality is my mental model of it. Hence why Packer & Nomad (and other tools) use `foo.hcl` named files and Terraform uses `foo.tf` named files.

I've pushed up a commit that adds HCL as a separate language.

Reply:
Reading the terraform-ls docs[^1] are _very_ clear on only using it for `*.tf`, `*.tfvar` files, not sure Zed can be configured to only sends some filetypes it supports through to the LSP. Also a bit confusing to a user that the "HCL" language is chosen and their `*.tf` files are linted/error checked with terraform-ls, but if they open a `config.pkr.hcl` it's also using "HCL" language but not linted/error checked with the LSP.

I've not yet found a standalone hcl-lsp so I suspect we'll have to do without for Packer, Nomad files for now (ie, HCL) and add terraform-ls for Terraform files.

[^1]: https://github.com/hashicorp/terraform-ls/blob/main/docs/USAGE.md#usage-of-terraform-language-server

Reply:
@dbanck might be able to chime in here. he works at Hashicorp and offered to help: https://github.com/zed-industries/zed/issues/5098#issuecomment-1911727160

Reply:
Thanks for the ping @matkam!

Looks like @caius and @fdionisi are on the right track. I think it makes sense to treat HCL and Terraform as different languages in Zed. This is how we do it for all other editors.

HCL is designed to be a toolkit for building languages, not a language itself, so it makes sense to provide only syntax highlighting via tree-stitter for it for now. There is some discussion about building a generic HCL formatter that applies basic formatting rules, but for now formatting is done by the applications that use HCL, such as `terraform fmt`, `packer fmt`, `vault policy fmt`, ...

For Terraform it makes sense to implement syntax highlighting via tree-sitter and language server support using https://github.com/hashicorp/terraform-ls. The language server can only handle Terraform-specific files (`.tf`, `.tf.json`, `.tfvars`, `.tfvars.json`, ...) and doesn't work well with generic `.hcl`. files. So it makes perfect sense to have separate language configurations.

There are no other language servers that I'm aware of, but if there is a Packer LS one day, it would need it's own language id and configuration (similar to Terraform today). This language configuration would claim the file extension `.pkr.hcl` and be more specific than the generic `.hcl` `path_suffixes`.